app/controllers/settings/ldap_test_controller.rb

```diff
@@ -0,0 +1,136 @@
+    render json: result_hash
+  end
+
+  # Returns an ldap object
+  def get_ldap(host, port, start_tls, pem_string, anon_bind, dN, pass)
+    tls = if start_tls == "true"
+      :start_tls
+    else
+      :simple_tls
+    end
+
+    decoded_pem = Base64.decode64(pem_string)
+    cert = OpenSSL::X509::Certificate.new(decoded_pem)
+    cert_store = OpenSSL::X509::Store.new
+    cert_store.add_cert(cert)
+    tls_options = {
+      verify_mode: OpenSSL::SSL::VERIFY_PEER,
+      cert_store:  cert_store
+    }
+
+    auth_hash = if anon_bind == "true"
+      {
+        method: :anonymous
+      }
+    else
+      {
+        method:   :simple,
+        username: dN,
+        password: pass
+      }
+    end
+    Net::LDAP.new host:       host,
+                  port:       port,
+                  auth:       auth_hash,
+                  encryption: {
+                    method:      tls, # could also be :start_tls
+                    tls_options: tls_options
+                  }
+  end
+
+  def create
+    result_hash = { initial_bind: {}, user_search: {}, result: {} }
+    result_hash.store("params", params)
+
+    # Incoming params
+    host = params[:host]
+    port = params[:port]
+    start_tls = params[:start_tls]
+    pem_string = params[:cert]
+    anon_bind = params[:anon_bind]
+    dN = params[:dN]
+    pass = params[:pass]
+    baseDN = params[:baseDN]
+    filter_string = params[:filter]
+
+    begin
+      ldap = get_ldap(host, port, start_tls, pem_string, anon_bind, dN, pass)
+    rescue OpenSSL::X509::CertificateError
+      result_hash.store("error", "Invalid Cerificate Format")
+      render_result(result_hash, "fail", "Invalid certificate, check format and try again")
+      return
+    # :nocov:
+    rescue StandardError => error
+      # Using StandardError because each OpenSSL::X509 errors must be caught individually
+      # and only CertificateError is documented: https://ruby-doc.org/stdlib-2.5.0/libdoc/openssl/rdoc/OpenSSL/X509/CertificateError.html
+      result_hash.store("error", error.inspect)
+      render_result(result_hash, "fail", "Unspecified certificate error.")
+      return
+    # :nocov:
+    end
+
+    # Initial bind with DN to test connection to LDAP server
+    begin
+      bind_result = ldap.bind
+      result_hash[:initial_bind] = {
+        result_code:    ldap.get_operation_result.code,
+        result_message: ldap.get_operation_result.message
+      }
+
+      unless bind_result
+        render_result(result_hash, "fail", "Failed to bind to LDAP server: " \
+        "#{ldap.get_operation_result.message}")
+        return
+      end
+    rescue Net::LDAP::Error => error
+      result_hash.store("error", error.inspect)
+      render_result(result_hash, "fail", "Check that host/port are correct and reachable.")
+      return
+    end
+
+    # Search for user
+    begin
+      user_search_records = ldap_filter_search(ldap, baseDN, filter_string)
+      if !user_search_records.empty?
```

Comment:
Consider reversing the if so that the '!' can be dropped

Reply:
This format was chosen as personal convention to to keep the passing condition at top and failing condition at bottom.